sparse/impl/KokkosSparse_spmv_impl_merge.hpp

```diff
@@ -0,0 +1,374 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOSSPARSE_SPMV_IMPL_MERGE_HPP
+#define KOKKOSSPARSE_SPMV_IMPL_MERGE_HPP
+
+#include <sstream>
+
+#include "KokkosKernels_Iota.hpp"
+#include "KokkosKernels_AlwaysFalse.hpp"
+
+#include "KokkosSparse_merge_matrix.hpp"
+
+namespace KokkosSparse::Impl {
+
+/*! \brief Merge-based SpMV
+  Hierarchical GPU implementation
+  Each team uses MergePath search to find the non-zeros and rows it is
+  responsible for Each thread in the team similarly uses diagonal search within
+  the team to determine which entries it will be responsible for
+  The threads then atomically accumulate partial produces
+*/
+template <class ExecutionSpace, class AMatrix, class XVector, class YVector>
+struct SpmvMergeHierarchical {
+  using device_type    = typename YVector::device_type;
+  using exec_space     = ExecutionSpace;
+  using y_value_type   = typename YVector::non_const_value_type;
+  using x_value_type   = typename XVector::non_const_value_type;
+  using A_value_type   = typename AMatrix::non_const_value_type;
+  using A_ordinal_type = typename AMatrix::non_const_ordinal_type;
+  using A_size_type    = typename AMatrix::non_const_size_type;
+  using row_map_non_const_value_type =
+      typename AMatrix::row_map_type::non_const_value_type;
+
+  using policy_type = Kokkos::TeamPolicy<exec_space>;
+  using team_member = typename policy_type::member_type;
+
+  using um_row_map_type =
+      Kokkos::View<typename AMatrix::row_map_type::data_type,
+                   typename AMatrix::row_map_type::device_type::memory_space,
+                   Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+  using row_map_scratch_type =
+      Kokkos::View<typename AMatrix::row_map_type::non_const_data_type,
+                   typename exec_space::scratch_memory_space,
+                   Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+  using iota_type = KokkosKernels::Impl::Iota<A_size_type, A_size_type>;
+
+  using DSR = typename KokkosSparse::Impl::MergeMatrixDiagonal<
+      um_row_map_type, iota_type>::position_type;
+
+  using KAT = Kokkos::ArithTraits<A_value_type>;
+
+  // results of a lower-bound and upper-bound diagonal search
+  struct Chunk {
+    DSR lb;  // lower bound
+    DSR ub;  // upper bound
+  };
+
+  template <bool NONZEROS_USE_SCRATCH, bool ROWENDS_USE_SCRATCH,
+            bool Y_USE_SCRATCH, bool CONJ>
+  struct Functor {
```

Comment:
I guess that's fine but in general this is the type of name that can easily clash with other struct in the same namespace... renaming it or putting it in an anonymous namespace might be better?